Bypasseable Errors for Rack Attack Proxy Instances

mpenalozag opened the pull request:
## Description
Some errors on proxy instances could lead to requests not being passed to the next middleware. This could be behaviour that we want to avoid, we wouldn't want our system to stop responding just because of unexpected issues between Rack::Attack and our external Cache instance.

## Considerations
- This is a short draft to validate the approach. Happy to iterate it.
- I noticed there could be more tests on some behaviours. Happy to add those after we agree on a solution

santib commented on lib/rack/attack/store_proxy/dalli_proxy.rb:
```diff
@@ -67,10 +67,10 @@ def with
-          nil
+        def should_bypass_error?(error)
+          # Dalli-specific default behavior: bypass Dalli errors
+          return true if defined?(::Dalli::DalliError) && error.is_a?(Dalli::DalliError)
```
I like the idea to keep the existing behavior untouched, but maybe we can do something else to not have to override the `should_bypass_error?` method here?

maybe something like having this DalliProxy have `bypassable_store_errors` default to `['Dalli::DalliError']` as a String so that it doesn't fail if the gem is not installed?

mpenalozag replied:
I think that's a good idea. Like you said, we could just define the `bypassable_store_errors` and let the `BaseProxy` class handle the rest.

santib commented on lib/rack/attack/base_proxy.rb:
```diff
@@ -5,6 +5,43 @@
     class BaseProxy < SimpleDelegator
+      attr_reader :bypass_all_store_errors, :bypassable_store_errors
+
+      def initialize(store, bypass_all_store_errors: false, bypassable_store_errors: [])
```
wondering if it's better to have just one variable: `bypassable_store_errors` which can either by a `Symbol` (:all, :none), or an `Array`? Just to avoid the possible combinations of both variable sets:
- bypass_all_store_errors: false, bypassable_store_errors: [...]
- bypass_all_store_errors: true, bypassable_store_errors: []
- bypass_all_store_errors: false, bypassable_store_errors: [...]
- bypass_all_store_errors: true, bypassable_store_errors: []

mpenalozag replied:
I think that's a good approach. We could also raise a configuration error if both variables are defined

santib commented:
Hey @mpenalozag, he may not be available at the time. So if you want we can go ahead with the PR 👍

grzuy commented on lib/rack/attack/cache.rb:
```diff
@@ -19,10 +19,10 @@ def initialize(store: self.class.default_store)
       attr_reader :store
 
-      def store=(store)
+      def store=(store, bypass_all_store_errors: false, bypassable_store_errors: [])
```
How would this be called from the caller? 
Thinking about what the end user API would be.

grzuy reviewed (commented):
Hey there!

Sorry for the delay.

@mpenalozag thanks for you contribution!

Interested and eager to go a bit deeper on the issue/problem being addressed here and what others (if any) alternatives exist to address it, before considering adding a new feature that changes the public API in some way.

Can you elaborate on which particular proxy you're using and which errors in particular are making this a problem for you?

Thank you in advance!

mpenalozag commented:
@grzuy @santib Hey guys! Doing a follow up on this. Should I just add the tests and wait for you review?

Commit:
feat: Introduce bypassable_store_errors option in base proxy

Replaces the bypass_all_store_errors + bypassable_store_errors pair with
a single option that accepts :all, :none, or an Array of Classes / String
class names (Strings resolve by ancestor-name match so missing gems don't
crash loading). Subclasses can declare their own default via
default_bypassable_store_errors. Invalid values raise
MisconfiguredStoreError.

Commit:
feat: Wire bypassable_store_errors through Cache

Adds Cache#bypassable_store_errors= as a separate accessor so callers
can configure the option without the awkward multi-arg setter pattern.
Both store= and bypassable_store_errors= re-wrap the underlying store,
so configuration order is irrelevant.

Commit:
feat: Route DalliProxy errors through handle_store_error

Replaces the bespoke rescuing helper with handle_store_error and
declares ['Dalli::DalliError'] as DalliProxy.default_bypassable_store_errors.
Because the entry is a String, loading DalliProxy does not require the
dalli gem to be present.

Commit:
feat: Route RedisProxy errors through handle_store_error

Replaces the bespoke rescuing helper with handle_store_error and
declares ['Redis::BaseConnectionError'] as the default bypass list.

Commit:
feat: Route remaining proxies through handle_store_error and document

Applies handle_store_error in RedisStoreProxy, RedisCacheStoreProxy and
MemCacheStoreProxy so their calls participate in the shared bypass
mechanism. Adds cross-proxy specs pinning each subclass default and
documents bypassable_store_errors in the README.

Commit:
Merge upstream/main into feat/bypasseable-errors

Resolves cache.rb conflict: preserves the wrap_store-based store=
setter (needed for bypassable_store_errors) and adds the eager
check_store_methods_presence call introduced in #714.
